game-headed/src/test/java/org/triplea/game/client/ui/javafx/screens/MapSelectionTest.java

```diff
@@ -0,0 +1,121 @@
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.properties.GameProperties;
+import games.strategy.engine.framework.ui.GameChooserEntry;
+import java.lang.reflect.Field;
+import java.util.List;
+import javafx.scene.Node;
+import javafx.scene.layout.Pane;
+import javafx.scene.web.WebEngine;
+import javafx.scene.web.WebView;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.triplea.game.client.parser.GameDetector;
+import org.triplea.game.client.ui.javafx.screen.ScreenController;
+import org.triplea.game.client.ui.javafx.util.FxmlManager;
+
+public class MapSelectionTest {
+  private final String mapName = "\n Test Name \n";
+  private final String mapNotes = "  \n  Test Notes  \r\n  ";
+
+  private GameDetector gameDetector;
+  private MapSelection mapSelection;
+
+  @BeforeEach
+  void setUp() {
+    gameDetector = mock(GameDetector.class);
+    mapSelection = new MapSelection(gameDetector, String::concat);
+  }
+
+  @Test
+  void onShowTriggersLoadingCorrectly() {
+    mapSelection.onShow();
+
+    verify(gameDetector).discoverGames(any());
+
+    clearInvocations(gameDetector);
+
+    mapSelection.onShow();
+
+    verify(gameDetector, never()).discoverGames(any());
+  }
+
+  @Test
+  void mapSelectionObjectReturnsCorrectNode() {
+    final Node node = mock(Node.class);
+    mapSelection.setRoot(node);
+
+    assertThat(mapSelection.getNode(), is(equalTo(node)));
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  void backKeyWorks() {
+    final ScreenController<FxmlManager> manager = mock(ScreenController.class);
+
+    mapSelection.connect(manager);
+
+    mapSelection.backToGameSelection();
+
+    verify(manager).switchScreen(FxmlManager.GAME_SELECTION_CONTROLS);
+  }
+
+  @Test
+  void closeButtonClosesDetails() {
+    final Node node = mock(Pane.class);
+    mapSelection.setPreviewContainer(node);
+
+    mapSelection.closeDetails();
+
+    verify(node).setVisible(false);
+  }
+
+  @Test
+  void detailsButtonShowsDetails() throws Exception {
+    final Node previewContainer = mock(Node.class);
+    mapSelection.setPreviewContainer(previewContainer);
+
+    final WebView previewWindow = mock(WebView.class);
+    final var webEngine = mockWebEngine();
+    when(previewWindow.getEngine()).thenReturn(webEngine);
+    mapSelection.setPreviewWindow(previewWindow);
+
+    final GameChooserEntry gameChooserEntry = mockGameChooserEntryWithNotes();
+    mapSelection.setSelectedGame(gameChooserEntry);
+
+    mapSelection.showDetails();
+
+    verify(previewContainer).setVisible(true);
+    verify(webEngine).loadContent(mapNotes.trim() + mapName);
+  }
+
+  private WebEngine mockWebEngine() throws Exception {
+    final Class<?> clazz = Class.forName("com.sun.glass.ui.Screen");
+    final Field screensField = clazz.getDeclaredField("screens");
```

Comment:
Using reflection to set a component of the UI framework seems to potentially be dangerous or maybe even evil. Can you explain why alternatives were not feasible? I wonder a bit if the javafx code is mixed with the UI flow code a bit too much and if the two can be more decoupled. If we can move logic a bit further away from JavaFX, we may be able to avoid tests that need JavaFX component knowledge.

Reply:
It's a little bit hacky.
The problem in this case is the static initializer block of WebEngine, which is obviously executed once we try to mock this class, so this reflective call avoids some logic that tries to determine the primary screen etc. to be used by instances of the class.

I mean we could try using a framework like https://github.com/TestFX/TestFX + https://github.com/TestFX/Monocle which avoids this problem by "correctly" setting up a JavaFX environment for the tests, but for now this hack seems just good enough for me

Reply:
Monocle could be something that could help, but the point I'm trying to convey is that I don't think we should have any/much logic at the javafx layer. IE: all interactions to JavaFX are probably best done as part of injected strategy objects. This way we have 'pure' UI code that wires into strategy objects that do everything else.

In other words, the code we are testing seems to be too close to the UI code and we do not have them well enough decoupled.

Reply:
Said in yet another way, we might possibly be creating some of the same mistakes as we had with the swing code, where the UI layer and the view handling are not decoupled. Monocle would probably be useful that the UI is rendered and invokes the backend layer we create.